EIPS/eip-7504.md

```diff
@@ -0,0 +1,421 @@
+     *
+     *  @param functionSelector    The 4 byte selector of the function.
+     *  @param functionSignature   Function signature as a string. E.g. "transfer(address,address,uint256)"
+     */
+    struct ExtensionFunction {
+        bytes4 functionSelector;
+        string functionSignature;
+    }
+
+    /**
+     *  @notice An interface to describe an extension.
+     *
+     *  @param metadata     The extension's metadata; it's name, metadata URI and implementation contract address.
+     *  @param functions    The functions that belong to the extension.
+     */
+    struct Extension {
+        ExtensionMetadata metadata;
+        ExtensionFunction[] functions;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                            View Functions
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev Returns all extensions of the Router.
+    function getAllExtensions() external view returns (Extension[] memory allExtensions);
+}
+```
+
+The return value of `getAllExtensions` includes all callable functions on the contract, grouped by the implementation contract where they are implemented.
+
+For a given callable function on the contract, the return value of `getImplementationForFunction` for its function selector **MUST** match the implementation contract expressed for it in the return value of `getAllExtensions`.
+
+### Concepts
+
+An “upgradeable smart contract” is actually two kinds of smart contracts considered together as one system:
+
+1. **Proxy** smart contract: The smart contract whose state/storage we’re concerned with.
+2. **Implementation** smart contract: A stateless smart contract that defines the logic for how the proxy smart contract’s state can be mutated.
+
+![diagram showing a proxy contract that always performs delegateCall on the same implementation contract.](../assets/eip-7504/proxy-with-single-impl.png)
+
+The job of a proxy contract is to forward any calls it receives to the implementation contract via `delegateCall`. As a shorthand — a proxy contract stores state, and always asks an implementation contract how to mutate its state (upon receiving a call).
+
+This proposal introduces a `Router` smart contract.
+
+![diagram showing a router contract that delegateCalls a specific implementation based on the function called.](../assets/eip-7504/router-many-impls.png)
+
+Instead of always delegateCall-ing the same implementation contract, a `Router` delegateCalls a particular implementation contract (i.e. “Extension”) for the particular function call it receives.
+
+A router stores a map from function selectors → to the implementation contract where the given function is implemented. “Upgrading a contract” now simply means updating what implementation contract a given function, or functions are mapped to.
+
+![upgrading a router contract means updating its function selector → to implementation contract map.](../assets/eip-7504/router-upgrades.png)
```

Comment:
This section doesn't really have any tangible specifications in it. It probably belongs in the rationale section.